src/Logging/LoggingTrait.php

```diff
@@ -0,0 +1,166 @@
+<?php
+/**
+ * Copyright 2024 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Auth\Logging;
+
+use Psr\Log\LogLevel;
+
+trait LoggingTrait
+{
+    /**
+     * @param LogEvent $event
+     */
+    private function logRequest(LogEvent $event): void
+    {
+        $debugEvent = [
+            'timestamp' => $event->timestamp,
+            'severity' => strtoupper(LogLevel::DEBUG),
+            'clientId' => $event->clientId,
+            'requestId' => $event->requestId ?? null,
+        ];
+
+        $debugEvent = array_filter($debugEvent, fn ($value) => !is_null($value));
+
+        $jsonPayload = [
+            'request.method' => $event->method,
+            'request.url' => $event->url,
+            'request.headers' => $event->headers,
+            'request.payload' => $event->payload,
+            'request.jwt' => $this->getJwtToken($event->headers ?? []),
+            'retryAttempt' => $event->retryAttempt
+        ];
+
+        // Remove null values
+        $debugEvent['jsonPayload'] = array_filter($jsonPayload, fn ($value) => !is_null($value));
+
+        $stringifiedEvent = json_encode($debugEvent);
+
+        // There was an error stringifying the event, return to not break execution
+        if ($stringifiedEvent === false) {
+            return;
+        }
+
+        $this->logger->debug($stringifiedEvent);
+    }
+
+    /**
+     * @param LogEvent $event
+     */
+    private function logResponse(LogEvent $event): void
+    {
+        $debugEvent = [
+            'timestamp' => $event->timestamp,
+            'severity' => strtoupper(LogLevel::DEBUG),
+            'clientId' => $event->clientId,
+            'requestId' => $event->requestId ?? null,
+            'jsonPayload' => [
+                'response.headers' => $event->headers,
+                'response.payload' => $event->payload,
+                'latency' => $event->latency,
+            ]
+        ];
+
+        // Remove null values
+        $debugEvent = array_filter($debugEvent, fn ($value) => !is_null($value));
+        $debugEvent['jsonPayload'] = array_filter(
+            $debugEvent['jsonPayload'],
+            fn ($value) => !is_null($value)
+        );
+
+        $stringifiedEvent = json_encode($debugEvent);
+
+        // There was an error stringifying the event, return to not break execution
+        if ($stringifiedEvent === false) {
```

Comment:
do we still want to try to log the infoEvent even if the debug event fails? for instance do this instead?

```php
if ($stringifiedEvent !== false) {
    $this->logger->debug($stringifiedEvent);
}
```